### PR TITLE
Netqueues support for vendor and model

### PR DESCRIPTION
**Pull request by yanirq**
The net field under a profile can now pass to tuned specific net devices according to  interfaceName,device and vendor IDs.
Internal change - pass values as a udev regex that tuned supports.

### CODE REVIEW

File: pkg/controller/performanceprofile/components/tuned/tuned.go [hunk 1]
```diff
@@ -126,18 +126,23 @@ func NewNodePerformance(assetsDir string, profile *performancev2.PerformanceProf
 
 		for _, device := range profile.Spec.Net.Devices {
 			devices = make([]string, 0)
-			if device.InterfaceName != nil {
-				devices = append(devices, *device.InterfaceName)
+			if device.DeviceID != nil {
+				devices = append(devices, "^ID_MODEL_ID="+*device.DeviceID)
```
Comment (MarSik) [inline pkg/controller/performanceprofile/components/tuned/tuned.go:130]: The sorting here matters right? Can you please add a comment so nobody will forget in the future?

File: pkg/controller/performanceprofile/components/tuned/tuned.go [hunk 2]
```diff
@@ -126,18 +126,23 @@ func NewNodePerformance(assetsDir string, profile *performancev2.PerformanceProf
...
 			}
 			if device.VendorID != nil {
-				devices = append(devices, *device.VendorID)
+				devices = append(devices, "^ID_VENDOR_ID="+*device.VendorID)
 			}
-			if device.DeviceID != nil {
-				devices = append(devices, *device.DeviceID)
+			if device.InterfaceName != nil {
+				deviceNameAmendedRegex := strings.Replace(*device.InterfaceName, "*", ".*", -1)
```
Comment (fromanirh) [inline pkg/controller/performanceprofile/components/tuned/tuned.go:136]: why is this needed again? e.g. why can't we just expect `foobar.*` directly?

File: pkg/controller/performanceprofile/components/tuned/tuned.go [hunk 3]
```diff
@@ -126,18 +126,23 @@ func NewNodePerformance(assetsDir string, profile *performancev2.PerformanceProf
...
+				if strings.HasPrefix(*device.InterfaceName, "!") {
+					devices = append(devices, "^INTERFACE_NAME="+"(?!"+deviceNameAmendedRegex+")")
+				} else {
+					devices = append(devices, "^INTERFACE_NAME="+deviceNameAmendedRegex)
+				}
 			}
-			devicesArgs := strings.Join(devices, ",")
+			devicesUdevRegex := "r'" + strings.Join(devices, `[\s\S]*`) + "'"
```
Comment (fromanirh) [inline pkg/controller/performanceprofile/components/tuned/tuned.go:143]: could you please add a comment about how the final regex should look like? regexes may be hard enough, having them built programmatically is even harder.

File: pkg/controller/performanceprofile/components/tuned/tuned_test.go [hunk 4]
```diff
@@ -176,8 +176,83 @@ var _ = Describe("Tuned", func() {
...
+							{
+								InterfaceName: &netDeviceName,
+							},
+						}}
+					manifest := getTunedManifest(profile)
+					reservedSet, err := cpuset.Parse(string(*profile.Spec.CPU.Reserved))
+					Expect(err).ToNot(HaveOccurred())
+					reserveCPUcount := reservedSet.Size()
+					//regex field should be: devices_udev_regex=r'^INTERFACE_NAME=eth.*''
+					devicesUdevRegex := "r'\\^INTERFACE_NAME=" + strings.Replace(netDeviceName, "*", "\\.\\*", -1) + "'"
```
Comment (fromanirh) [inline pkg/controller/performanceprofile/components/tuned/tuned_test.go:193]: this is a bit harder to read than it could. Tests should be optimized for clarity, so I'd just write at the beginning of the block
```
netDeviceName := "eth0*"
devicesUdevRegex := "r\\'^INTERFACE_NAME=eth.*'
```

File: pkg/controller/performanceprofile/components/tuned/tuned_test.go [hunk 5]
```diff
@@ -176,8 +176,83 @@ var _ = Describe("Tuned", func() {
...
+					channelsRegex := regexp.MustCompile(`\s*\[net_1\]\\ntype=net\\ndevices_udev_regex=` + devicesUdevRegex + `\\nchannels=combined ` + strconv.Itoa(reserveCPUcount) + `\s*`)
```
Comment (fromanirh) [inline pkg/controller/performanceprofile/components/tuned/tuned_test.go:194]: same, I'd actually move this as a constant at the beginning of the block, because having a clear tests trumps some repetition

File: pkg/controller/performanceprofile/components/tuned/tuned_test.go [hunk 6]
```diff
@@ -176,8 +176,83 @@ var _ = Describe("Tuned", func() {
...
+					Expect(channelsRegex.MatchString(manifest)).To(BeTrue())
+				})
+				It("should set by negative interface name with reserved CPUs count", func() {
+					netDeviceName := "!ens5"
```
Comment (fromanirh) [inline pkg/controller/performanceprofile/components/tuned/tuned_test.go:198]: does this name depend on the host details on which we run?

File: pkg/controller/performanceprofile/components/tuned/tuned_test.go [hunk 7]
```diff
@@ -176,8 +176,83 @@ var _ = Describe("Tuned", func() {
...
+						Devices: []performancev2.Device{
+							{
+								InterfaceName: &netDeviceName,
+							},
+						}}
+					manifest := getTunedManifest(profile)
+					reservedSet, err := cpuset.Parse(string(*profile.Spec.CPU.Reserved))
+					Expect(err).ToNot(HaveOccurred())
+					reserveCPUcount := reservedSet.Size()
+					//regex field should be: devices_udev_regex=r'^INTERFACE_NAME=(?!ens5)'
```
Comment (fromanirh) [inline pkg/controller/performanceprofile/components/tuned/tuned_test.go:210]: same comments as per previous block

**Review by fromanirh — COMMENTED** [review]
mostly question/comments about maintnability

File: pkg/controller/performanceprofile/components/tuned/tuned.go (hunk 2 above)
Comment (yanirq) [reply pkg/controller/performanceprofile/components/tuned/tuned.go:136]: The intended wildcards to use in PAO are * and !  , I need to convert these under the hood so they match a python regular expression format

File: pkg/controller/performanceprofile/components/tuned/tuned_test.go (hunk 6 above)
Comment (yanirq) [reply pkg/controller/performanceprofile/components/tuned/tuned_test.go:198]: this needs to be tested e2e wise as well

File: pkg/controller/performanceprofile/components/tuned/tuned.go (hunk 1 above)
Comment (yanirq) [reply pkg/controller/performanceprofile/components/tuned/tuned.go:130]: ill add a comment. unittests should ensure this as well

File: pkg/controller/performanceprofile/components/tuned/tuned_test.go [hunk 8]
```diff
@@ -178,6 +178,9 @@ var _ = Describe("Tuned", func() {
 				})
 				It("should set by interface name with reserved CPUs count", func() {
 					netDeviceName := "eth*"
+					//regex field should be: devices_udev_regex=r'^INTERFACE_NAME=eth.*''
+					devicesUdevRegex := "r'\\^INTERFACE_NAME=" + strings.Replace(netDeviceName, "*", "\\.\\*", -1) + "'"
```
Comment (fromanirh) [inline pkg/controller/performanceprofile/components/tuned/tuned_test.go:182]: I meant to make this a string literal but no big deal, not worth a respin just for this.

**Review by fromanirh — APPROVED** [review]
/lgtm
/hold
giving time to other reviewers to chime in

File: pkg/controller/performanceprofile/components/tuned/tuned_test.go [hunk 9]
```diff
@@ -176,8 +176,83 @@ var _ = Describe("Tuned", func() {
...
+					netDeviceName := "eth*"
+					profile.Spec.Net = &performancev2.Net{
+						UserLevelNetworking: pointer.BoolPtr(true),
+						Devices: []performancev2.Device{
+							{
+								InterfaceName: &netDeviceName,
+							},
+						}}
+					manifest := getTunedManifest(profile)
+					reservedSet, err := cpuset.Parse(string(*profile.Spec.CPU.Reserved))
```
Comment (marcel-apf) [inline pkg/controller/performanceprofile/components/tuned/tuned_test.go:192]: are we sure the Spec.CPU.Reserved is not null?

File: pkg/controller/performanceprofile/components/tuned/tuned_test.go [hunk 10]
```diff
@@ -176,8 +176,83 @@ var _ = Describe("Tuned", func() {
...
+					netDeviceName := "!ens5"
+					profile.Spec.Net = &performancev2.Net{
+						UserLevelNetworking: pointer.BoolPtr(true),
+						Devices: []performancev2.Device{
+							{
+								InterfaceName: &netDeviceName,
+							},
+						}}
+					manifest := getTunedManifest(profile)
+					reservedSet, err := cpuset.Parse(string(*profile.Spec.CPU.Reserved))
```
Comment (marcel-apf) [inline pkg/controller/performanceprofile/components/tuned/tuned_test.go:211]: are we sure the Spec.CPU.Reserved is not null?

File: pkg/controller/performanceprofile/components/tuned/tuned_test.go [hunk 11]
```diff
@@ -176,8 +176,83 @@ var _ = Describe("Tuned", func() {
...
+					netDeviceVendorID := "0x1af4"
+					profile.Spec.Net = &performancev2.Net{
+						UserLevelNetworking: pointer.BoolPtr(true),
+						Devices: []performancev2.Device{
+							{
+								VendorID: &netDeviceVendorID,
+							},
+						}}
+					manifest := getTunedManifest(profile)
+					reservedSet, err := cpuset.Parse(string(*profile.Spec.CPU.Reserved))
```
Comment (marcel-apf) [inline pkg/controller/performanceprofile/components/tuned/tuned_test.go:230]: same

File: pkg/controller/performanceprofile/components/tuned/tuned_test.go [hunk 12]
```diff
@@ -176,8 +176,83 @@ var _ = Describe("Tuned", func() {
...
+					profile.Spec.Net = &performancev2.Net{
+						UserLevelNetworking: pointer.BoolPtr(true),
+						Devices: []performancev2.Device{
+							{
+								DeviceID: &netDeviceModelID,
+								VendorID: &netDeviceVendorID,
+							},
+						}}
+					manifest := getTunedManifest(profile)
+					reservedSet, err := cpuset.Parse(string(*profile.Spec.CPU.Reserved))
```
Comment (marcel-apf) [inline pkg/controller/performanceprofile/components/tuned/tuned_test.go:251]: same

File: pkg/controller/performanceprofile/components/tuned/tuned_test.go [hunk 13]
```diff
@@ -186,15 +261,40 @@ var _ = Describe("Tuned", func() {
 						Devices: []performancev2.Device{
 							{
 								InterfaceName: &netDeviceName,
+								DeviceID:      &netDeviceModelID,
 								VendorID:      &netDeviceVendorID,
+							},
+						}}
+					manifest := getTunedManifest(profile)
+					reservedSet, err := cpuset.Parse(string(*profile.Spec.CPU.Reserved))
```
Comment (marcel-apf) [inline pkg/controller/performanceprofile/components/tuned/tuned_test.go:274]: same

File: pkg/controller/performanceprofile/components/tuned/tuned_test.go [hunk 14]
```diff
@@ -186,15 +261,40 @@ var _ = Describe("Tuned", func() {
...
+						UserLevelNetworking: pointer.BoolPtr(true),
+						Devices: []performancev2.Device{
+							{
+								InterfaceName: &netDeviceName,
 								DeviceID:      &netDeviceModelID,
+								VendorID:      &netDeviceVendorID,
 							},
 						}}
 					manifest := getTunedManifest(profile)
 					reservedSet, err := cpuset.Parse(string(*profile.Spec.CPU.Reserved))
```
Comment (marcel-apf) [inline pkg/controller/performanceprofile/components/tuned/tuned_test.go:297]: same

**Review by marcel-apf — COMMENTED** [review]
only a question - several times

File: pkg/controller/performanceprofile/components/tuned/tuned_test.go (hunk 9 above)
Comment (yanirq) [reply pkg/controller/performanceprofile/components/tuned/tuned_test.go:192]: the validation web hook will fail the profile creations if its null

File: pkg/controller/performanceprofile/components/tuned/tuned_test.go (hunk 10 above)
Comment (yanirq) [reply pkg/controller/performanceprofile/components/tuned/tuned_test.go:211]: same answer as above (validation web hook will fail creation of profile)

File: pkg/controller/performanceprofile/components/tuned/tuned_test.go (hunk 11 above)
Comment (yanirq) [reply pkg/controller/performanceprofile/components/tuned/tuned_test.go:230]: same answer as above (validation web hook will fail creation of profile)

File: pkg/controller/performanceprofile/components/tuned/tuned_test.go (hunk 12 above)
Comment (yanirq) [reply pkg/controller/performanceprofile/components/tuned/tuned_test.go:251]: same answer as above (validation web hook will fail creation of profile)

File: pkg/controller/performanceprofile/components/tuned/tuned_test.go (hunk 13 above)
Comment (yanirq) [reply pkg/controller/performanceprofile/components/tuned/tuned_test.go:274]: same answer as above (validation web hook will fail creation of profile)

File: pkg/controller/performanceprofile/components/tuned/tuned_test.go (hunk 9 above)
Comment (marcel-apf) [reply pkg/controller/performanceprofile/components/tuned/tuned_test.go:192]: great, thanks
